### PR TITLE
[#80] Quarkus OpenTracing reactive GRPC

### PR DESCRIPTION
**Pull request by mjurc**
* Migrating 010-quarkus-opentracing-reactive-grpc from beefy-scenarios

### CODE REVIEW

File: opentracing-reactive-grpc/src/test/java/io/quarkus/qe/containers/JaegerTestResource.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+import java.util.Map;
+import java.util.Optional;
+
+import org.eclipse.microprofile.config.inject.ConfigProperty;
+import org.testcontainers.containers.GenericContainer;
+import org.testcontainers.containers.wait.strategy.LogMessageWaitStrategy;
+
+import io.quarkus.test.common.QuarkusTestResourceLifecycleManager;
+
+public class JaegerTestResource implements QuarkusTestResourceLifecycleManager {
```
Comment (Sgitario) [inline opentracing-reactive-grpc/src/test/java/io/quarkus/qe/containers/JaegerTestResource.java:14]: This resource is no longer used.

File: opentracing-reactive-grpc/src/test/java/io/quarkus/qe/AbstractPingPongResourceIT.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    private static final String PONG_ENDPOINT = "/%s-pong";
+
+    private static final String PING_RESOURCE = "PingResource";
+    private static final String PONG_RESOURCE = "PongResource";
+
+    private static final int TRACE_PORT = 14268;
+    private static final int REST_PORT = 16686;
+
+    @ConfigProperty(name = JaegerTestResource.JAEGER_API_ENDPOINT, defaultValue = "http://localhost:16686/api/traces")
+    String jaegerEndpoint;
```
Comment (Sgitario) [inline opentracing-reactive-grpc/src/test/java/io/quarkus/qe/AbstractPingPongResourceIT.java:33]: This is not used.

File: opentracing-reactive-grpc/src/test/java/io/quarkus/qe/AbstractPingPongResourceIT.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+    @JaegerContainer(image = "jaegertracing/all-in-one:latest", tracePort = TRACE_PORT, restPort = REST_PORT, expectedLog = "\"Health Check state change\",\"status\":\"ready\"")
```
Comment (Sgitario) [inline opentracing-reactive-grpc/src/test/java/io/quarkus/qe/AbstractPingPongResourceIT.java:35]: You don't need to specify the image, tracePort, restPort and expectedLog, using simply `@JaegerContainer` works fine: if using ` .withProperty("quarkus.jaeger.endpoint", jaeger::getRestUrl);` and `.when().get(jaeger.getTraceUrl() + "?traceID=" + pingTraceId)`

File: opentracing-reactive-grpc/src/test/java/io/quarkus/qe/AbstractPingPongResourceIT.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    static final JaegerService jaeger = new JaegerService();
+
+    @QuarkusApplication
+    static RestService app = new RestService()
+            .withProperty("quarkus.jaeger.endpoint", jaeger::getTraceUrl);
```
Comment (Sgitario) [inline opentracing-reactive-grpc/src/test/java/io/quarkus/qe/AbstractPingPongResourceIT.java:40]: it should be `jaeger::getRestUrl`

File: opentracing-reactive-grpc/src/test/java/io/quarkus/qe/AbstractPingPongResourceIT.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        // Then both ping and pong rest endpoints should have the same trace Id.
+        String pingTraceId = given()
+                .when().get(pingEndpoint() + "/lastTraceId")
+                .then().statusCode(HttpStatus.SC_OK).and().extract().asString();
+
+        assertTraceIdWithPongService(pingTraceId);
+
+        // Then Jaeger is invoked
+        await().atMost(5, TimeUnit.SECONDS).untilAsserted(() -> given()
+                .when().get(jaeger.getRestUrl() + "?traceID=" + pingTraceId)
```
Comment (Sgitario) [inline opentracing-reactive-grpc/src/test/java/io/quarkus/qe/AbstractPingPongResourceIT.java:59]: it should be `jaeger.getTraceUrl()`

File: opentracing-reactive-grpc/src/test/java/io/quarkus/qe/GrpcPingPongResourceTest.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+import io.quarkus.test.scenarios.annotations.DisabledOnNative;
+
+@Disabled("Caused by https://github.com/quarkusio/quarkus/issues/13224")
+@QuarkusScenario
+@DisabledOnNative
+public class GrpcPingPongResourceTest extends AbstractPingPongResourceIT {
+
+    @Inject
+    @GrpcService
+    GrpcPongService pongService;
```
Comment (Sgitario) [inline opentracing-reactive-grpc/src/test/java/io/quarkus/qe/GrpcPingPongResourceTest.java:21]: This is not working using `@QuarkusScenario`, the workaround is to expose the last trace ID via a REST endpoint.

**Review by Sgitario — CHANGES_REQUESTED** [review]

**Review by pjgg — CHANGES_REQUESTED** [review]
On my local environment I am getting a compilation error:
cmd: `mvn clean verify`
Error:

```
Exception in thread "main" java.lang.AssertionError
        at jdk.compiler/com.sun.tools.javac.util.Assert.error(Assert.java:155)
        at jdk.compiler/com.sun.tools.javac.util.Assert.check(Assert.java:46)
        at jdk.compiler/com.sun.tools.javac.comp.Modules.enter(Modules.java:247)
        at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.readSourceFile(JavaCompiler.java:837)
        at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment$ImplicitCompleter.complete(JavacProcessingEnvironment.java:1530)
        at jdk.compiler/com.sun.tools.javac.code.Symbol.complete(Symbol.java:642)
        at jdk.compiler/com.sun.tools.javac.code.Symbol$ClassSymbol.complete(Symbol.java:1326)
        at jdk.compiler/com.sun.tools.javac.code.Type$ClassType.complete(Type.java:1140)
        at jdk.compiler/com.sun.tools.javac.code.Type$ClassType.getTypeArguments(Type.java:1066)
        at jdk.compiler/com.sun.tools.javac.code.Printer.visitClassType(Printer.java:237)
        at jdk.compiler/com.sun.tools.javac.code.Printer.visitClassType(Printer.java:52)
        at jdk.compiler/com.sun.tools.javac.code.Type$ClassType.accept(Type.java:993)
        at jdk.compiler/com.sun.tools.javac.code.Printer.visit(Printer.java:136)
        at jdk.compiler/com.sun.tools.javac.util.AbstractDiagnosticFormatter.formatArgument(AbstractDiagnosticFormatter.java:199)


```

File: opentracing-reactive-grpc/pom.xml [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>io.quarkus</groupId>
+                <artifactId>quarkus-maven-plugin</artifactId>
+                <executions>
+                    <execution>
+                        <goals>
+                            <goal>generate-code</goal>
+                            <goal>generate-code-tests</goal>
```
[the remark this answers is not in the document]
Comment (mjurc) [reply opentracing-reactive-grpc/pom.xml]: Not sure if there would be any benefit. What would that help?

File: opentracing-reactive-grpc/pom.xml [hunk 8]
```diff
@@ -0,0 +1,84 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>io.quarkus.ts.qe</groupId>
+        <artifactId>parent</artifactId>
+        <version>1.0.0-SNAPSHOT</version>
+        <relativePath>..</relativePath>
+    </parent>
+    <artifactId>opentracing-reactive-grpc</artifactId>
```
[the remark this answers is not in the document]
Comment (mjurc) [reply opentracing-reactive-grpc/pom.xml]: Moved.

File: opentracing-reactive-grpc/src/test/java/io/quarkus/qe/GrpcPingPongResourceTest.java (hunk 6 above)
Comment (mjurc) [reply opentracing-reactive-grpc/src/test/java/io/quarkus/qe/GrpcPingPongResourceTest.java]: Right, I've updated the test. Can you double check if this is suitable now?

**Review by Sgitario — APPROVED** [review]

**Comment by Sgitario** [comment]
@mjurc merged, please blackport PR to 2.2

**Pull request merged**